contracts/golos.worker/main.cpp

```diff
@@ -0,0 +1,1088 @@
+
+    get_proposals().modify(proposal_ptr, proposal_ptr->tspec_author, [&](proposal_t &proposal) {
+      proposal.tspec.update(data);
+    });
+  }
+
+  /**
+   * @brief startwork chooses worker account and allows the worker to start work on the proposal
+   * @param proposal_id proposal ID
+   * @param worker worker account name
+   */
+  /// @abi action
+  void startwork(proposal_id_t proposal_id, account_name worker)
+  {  
+    LOG("proposal_id: %, worker: %", proposal_id, ACCOUNT_NAME_CSTR(worker));
+    auto proposal_ptr = get_proposal(proposal_id);
+    eosio_assert(proposal_ptr->state == proposal_t::STATE_TSPEC_CREATE, "invalid proposal state");
+    eosio_assert(proposal_ptr->type == proposal_t::TYPE_1, "unsupported action");
+    require_auth(proposal_ptr->tspec_author);
+
+    get_proposals().modify(proposal_ptr, proposal_ptr->tspec_author, [&](proposal_t &proposal) {
+      proposal.worker = worker;
+      proposal.work_begining_time = TIMESTAMP_NOW;
+      proposal.set_state(proposal_t::STATE_WORK);
+    });
+  }
+
+  /**
+   * @brief cancelwork cancels work. Can be called by the worker or technical specification author
+   * @param proposal_id propsal ID
+   * @param initiator a cancel initiator's account name
+   */
+  /// @abi action
+  void cancelwork(proposal_id_t proposal_id, account_name initiator)
+  {
+    LOG("proposal_id: %, initiator: %", proposal_id, ACCOUNT_NAME_CSTR(initiator));
+    auto proposal_ptr = get_proposal(proposal_id);
+    eosio_assert(proposal_ptr->state == proposal_t::STATE_WORK, "invalid proposal state");
+    eosio_assert(proposal_ptr->type == proposal_t::TYPE_1, "unsupported action");
+
+    if (initiator == proposal_ptr->worker)
+    {
+      require_auth(proposal_ptr->worker);
+    }
+    else
+    {
+      require_auth(proposal_ptr->tspec_author);
+    }
+
+    get_proposals().modify(proposal_ptr, initiator, [&](proposal_t &proposal) {
+      refund(proposal, initiator);
+    });
+  }
+
+  /**
+   * @brief poststatus post status for the work done
+   * @param proposal_id proposal ID
+   * @param comment_id comment ID
+   * @param comment comment data
+   * @param finished true if all work done
+   */
+  /// @abi action
+  void poststatus(proposal_id_t proposal_id, comment_id_t comment_id, const comment_data_t &comment, bool finished)
+  {
+    LOG("proposal_id: %, comment: %, final: %", proposal_id, comment.text.c_str(), (int) finished);
+    auto proposal_ptr = get_proposal(proposal_id);
+    eosio_assert(proposal_ptr->state == proposal_t::STATE_WORK, "invalid proposal state");
+    eosio_assert(proposal_ptr->type == proposal_t::TYPE_1, "unsupported action");
+    require_auth(proposal_ptr->worker);
+
+    get_proposals().modify(proposal_ptr, proposal_ptr->worker, [&](auto &proposal) {
+      proposal.work_status.add(comment_id, proposal.worker, comment);
+
+      if (finished)
+      {
+        proposal.set_state(proposal_t::STATE_TSPEC_AUTHOR_REVIEW);
```

Comment:
this stage is not needed, it's off-chain. worker and tspec author communicate, so here is no reason to worker to mark work finished. tspec author knows when work is finished and initiates "delegates review" from `STATE_WORK` state

Reply:
A flag `finished` was discussed in the comment in the software requirements specification in google docs. Ok, I removed it.

Reply:
https://github.com/GolosChain/golos.worker/pull/1/commits/1a161585c2f0e3b0a033b47f899e09e0bb57b6b7 fixes this partly, please, update js too (wrong state enum, etc) and fix https://github.com/GolosChain/golos.worker/pull/1#discussion_r231797352